eng/pipelines/ci.yml

```diff
@@ -67,30 +67,19 @@ jobs:
     helixQueueGroup: ci
     jobParameters:
       testGroup: outerloop
-
-#
-# ReadyToRun test jobs
-#
-- template: /eng/platform-matrix.yml
-  parameters:
-    jobTemplate: test-job.yml
-    buildConfig: checked
-    platforms:
-    - Linux_arm64
-    - Linux_musl_x64
-    - Linux_musl_arm64
-    - Linux_rhel6_x64
-    - Linux_x64
-    - OSX_x64
-    - Windows_NT_x64
-    - Windows_NT_x86
-    - Windows_NT_arm
-    - Windows_NT_arm64
-    helixQueueGroup: ci
-    jobParameters:
-      testGroup: outerloop
-      readyToRun: true
-      displayNameArgs: R2R
+      jitPlatforms:
```

Comment:
@trylek Let me suggest a slightly different way of what you are trying to accomplish here. 

Can we stop using *combined* `test-job.yml` in the pipeline files and instead use `build-test-job.yml` and `run-test-job.yml` directly? We can do this only for complex pipelines like ci.yml or internal.yml.

For example, `ci.yml` would look like
```yaml
- template: /eng/platform-matrix.yml
  parameters:
    jobTemplate: build-test-job.yml
    buildConfig: checked
    platformGroup: all
    jobParameters:
      testGroup: outerloop

- template: /eng/platform-matrix.yml
  parameters:
    jobTemplate: run-test-job.yml
    buildConfig: checked
    platformGroup: all
    helixQueueGroup: ci
    jobParameters:
      testGroup: outerloop

- template: /eng/platform-matrix.yml
  parameters:
    jobTemplate: run-test-job.yml
    buildConfig: checked
    helixQueueGroup: ci
    platforms:
    - Linux_arm64
    - <Other R2R Platforms>
    jobParameters:
      testGroup: outerloop
      readyToRun: true
```

The reasoning behind this:

`platform-matrix.yml` was originally designed to not have any dependencies on the job it tries to *"multiply"*. In other words, `platform-matrix.yml` acts as a "high order template" that *"multiplies*" `jobTemplate`-jobs given `platformGroup`/`platforms` and pass `jobParameters` as a opaque dictionary argument. I think this is important for keeping `platform-matrix.yml` universally re-usable

Reply:
I like this idea, thanks Egor, I'll do that. In fact it was probably just due to my limited understanding that I didn't make the split right away when I was splitting test-job into build-test-job & run-test-job.

Reply:
Hmm, actually there's a subtle problem to this approach regarding several flavors of code duplication. One thing is that splitting build-test-job and run-test-job will typically require us duplicating the platform list. Another is that today the build OS selection logic is based exactly on the assumption that the platform-matrix receives the "entire" matrix - in your example snippet, the filtered list of platforms for the R2R version of run-test-job basically means that platform-matrix has no way of knowing that build-test-job ran for a broader set of platforms it can pick up managed artifacts from. Fixing this seems to require explicitly specifying the OS for managed component build for each of the platforms. One way to consolidate this with your suggestion might be to just create a "customized version of platform-matrix for test-job". I'm going to spend a bit more time trying to figure out whether we might be able to do something more elegant here.

Reply:
As a 2nd attempt I've done what I described above - I have created a new script "test-platform-matrix" as a customized version of "platform-matrix" for "test-job". Self-reviewing the code I think it simplified things overall and it also let me remove some remaining duplication in test-job.

Reply:
@trylek

> One thing is that splitting build-test-job and run-test-job will typically require us duplicating the platform list. 

 I don't think we should worry about the code duplication much. The split should only be needed for pipelines that have both non-R2R and R2R jobs - ci.yml pr.yml and internal.yml. For other pipelines we should continue to use the test-job.yml. With your approach you would still need the duplication of platforms when you define jitPlatforms and r2rPlatforms.


> in your example snippet, the filtered list of platforms for the R2R version of run-test-job basically means that platform-matrix has no way of knowing that build-test-job ran for a broader set of platforms it can pick up managed artifacts from.

I am probably missing something here. How run-test-job find the managed artifacts currently?

Reply:
The logic I recently added resides in the three-stage split of the platform-matrix script (platform-matrix / platform-matrix-combos / platform-matrix-managed-test-build): platform-matrix first evaluates all platform flavors to run and passes this list to platform-matrix-combos; this second script then uses the full platform list to pick a suitable ("the fastest") OS to build managed artifacts for each test run on; the third script (platform-matrix-managed-test-build) then passes this information back to test-job.yml which uses it to filter out which test build jobs should actually run. I know it may sound complicated but so far I'm not aware of any more compact way to represent this.

Reply:
I see. Let me check that I've got this right. Basically, we build tests for each win-$(arch) combo and for each $(arch) in case of non-Windows (i.e. we assume that OSX and Linux managed tests assets are interchangeable) ?

Reply:
Yes, I believe that to be correct. According to Jarret, after switching over to building against reference assemblies we should be ultimately able to remove even the Windows vs. Linux+OSX dichotomy and build all managed test artifacts on OSX x64, that should be fastest. We're not yet able to do this and we cannot yet drop the dependency of managed test artifacts on targeting architecture as the existing CoreCLR scripts invoke Roslyn in such a way that it stores the architecture ID in the MSIL output file and I haven't yet figured out how to fix this.

Reply:
Before my initial change splitting the build of managed and native test artifacts, PR runs had 16 test runs so they implicitly executed 16 test build runs. My initial change,

https://github.com/dotnet/coreclr/pull/26581

reduced the number of test build jobs from 16 to 13. This change further reduces it to 8; after fixing the architecture dependency, we should be down to 4 (Windows / OSX checked / release) and after removing the remaining OS dependency we should be at 2 which is the hard limit (OSX checked / release).

Reply:
In fact, today even the output R2R PE executables are identical between Linux and OSX (for the same architecture) even though this may change in the future as there are several motivations for switching over to OS-native executable formats.

Reply:
@trylek Thanks for explaining this! 

To address your previous statement:

>  in your example snippet, the filtered list of platforms for the R2R version of run-test-job basically means that platform-matrix has no way of knowing that build-test-job ran for a broader set of platforms it can pick up managed artifacts from.

I think it's fine and we can safely assume in run-test-job that the corresponding build-test-job for given combo of `os`x`arch`x`subarch`x`flavor` was defined in the corresponding without explicit passing of this knowledge down to run-test-job. If this assumptions is broken this can easily be seen/tested during the pipeline run.

Reply:
Well, I'm afraid the problem is deeper than that. At the very least my initial implementation tried to be "lossless" in the sense of catering for arbitrary combination of tests to run. This forced me to coin the above described complex logic for choosing the appropriate managed build platform among an arbitrary set of overall platforms to test.

I'm not necessarily saying this is a sacred invariant, it was just something I treated as an invariant. For instance, we can drop part of the generality by hard-coding the logic for choosing the OS to build managed test assets on (e.g. "always build managed components for any Linux flavor on OSX"). This would simplify the test scripts at the expense of some duplication.

In fact, this would let me roll back the three-stage platform-matrix script split as the "managedBuild*" parameters passed to the template being expanded would become simple constants. This would only become a real concern if we decided to let developers trigger targeted pipelines limited to a particular platform in which case we would need to implement the matching logic selecting the OS for managed asset build.

Reply:
In my opinion, *"always build managed components for any Linux flavor on OSX"* is alright simplification.

But I don't want to push on what I am suggesting. I am just trying to say - if we can keep the CI system simpler - great. And making assumptions like those seems reasonable in order to achieve this. If we can not and the added complexity is justified by short or long term benefits - that's also fine with me.

Reply:
Thanks Egor for your feedback - I do value it, that's why I asked for it and that's why I'm trying to figure out how to ship it, so to say. Frankly speaking I would love to ditch "test-job.yml" altogether because two ways of defining pipelines mean two ways of possible failures. Short-circuiting most Linux managed test builds to use OSX-built artifacts is also something that seems trivial to ensure. In fact, the only downside I'm fighting right now is restoration of the ugly conditional statement composing the build / run test names. If only the Azure Pipelines yml syntax was designed by someone familiar with programming languages :-)!